app/src/modules/whyis-dataset.js

```diff
@@ -0,0 +1,279 @@
+      } else if (Object.getOwnPropertyDescriptor(datasetFieldUris, val.toLowerCase())) { // Recursive case (val is an allowed field)
+        fieldDict[datasetFieldUris[val.toLowerCase()]] = recursiveFieldSetter([datasetFieldUris[val.toLowerCase()], value[val]])
+      } else {
+        fieldDict['@value'] = value
+      }
+    }
+    return fieldDict
+  }
+}
+
+// Blank dataset
+function getDefaultDataset () {
+  return Object.assign({}, defaultDataset)
+}
+
+function deleteDataset (datasetUri) {
+  return listNanopubs(datasetUri)
+    .then(nanopubs => {
+      Promise.all(nanopubs.map(nanopub => deleteNanopub(nanopub.np)))
+    }
+    )
+}
+
+// Handle all of the uploads as multipart form
+async function saveDataset (dataset, fileList, image, guuid) {
+  let p = Promise.resolve()
+  if (dataset.uri) {
+    p = deleteDataset(dataset.uri)
+  } else if (arguments.length === 1) {
+    dataset.uri = generateDatasetId()
+  } else {
+    dataset.uri = generateDatasetId(guuid)
+  }
+  const [distrRes, imgRes] = await Promise.all([
+    saveDatasetFiles(fileList),
+    saveDatasetFiles([image]), // TODO: verify we want to use the same function for depiction
+    p
+  ])
+  const datasetLd = buildDatasetLd(dataset)
+  if (distrRes?.files?.length) {
+    datasetLd[datasetFieldUris.distribution] = buildDistrLd(distrRes?.files)
+  } if (imgRes?.files?.length) {
+    datasetLd[datasetFieldUris.depiction] = buildDepictionLd(imgRes?.files?.[0], dataset.uri)
+  } return postNewNanopub(datasetLd)
+  // TODO: Error handling
+}
+
+async function saveDatasetFiles (fileList) {
+  if (fileList.length) {
+    const url = `${window.location.origin}/api/files/upload`
+    const formData = new FormData()
+    fileList.forEach((file) => formData.append('uploadfile', file?.file ?? file))
+    const result = await fetch(url, {
+      method: 'POST',
+      body: formData,
+      redirect: 'follow',
+      headers: {
+        Authorization: store.getters['auth/token']
+      }
+    })
+    return await result.json()
+    // TODO: Error handling
+  }
+}
+
+function buildDistrLd (fileList) {
+  const distrLDs = Array(fileList.length)
+  Array
+    .from(Array(fileList.length).keys())
+    .map(x => {
+      // TODO: check if we want to keep distribution uri as /explorer/dataset/id/filename and redirect for download
+      distrLDs[x] = {
+        '@id': `${window.location.origin}/api/files?filename=${fileList[x].filename}`,
+        '@type': 'http://purl.org/net/provenance/ns#File',
+        'http://www.w3.org/2000/01/rdf-schema#label': fileList[x].originalname
+      }
+    })
+  return distrLDs
+}
+
+function buildDepictionLd (file, uri) {
+  const depictionLd = {
+    '@id': `${uri}/depiction`,
+    '@type': 'http://purl.org/net/provenance/ns#File',
+    'http://www.w3.org/2000/01/rdf-schema#label': file.originalname,
+    'http://w3.org/ns/dcat#accessURL': `${window.location.origin}/api/files?filename=${file.filename}`
```

Comment:
`${window.location.origin}/api/files?filename=${fileList[x].filename}`.

Reply:
For this one did you mean `${window.location.origin}/api/files/get?filename=${file.filename}` ?